File: sample.cs
// Time Complexity : O(m*n)
// Space Complexity :O(Min(m,n))
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : yes. Syntax errors and language c# is different than java
public class Solution {
    public int NumIslands(char[][] grid) {
        int m = grid.Length;
        int n = grid[0].Length;
        int count = 0;
        int[][] dirs = new int[][]{new int[]{-1,0}, new int[]{1,0}, new int[]{0,1}, new int[]{0,-1}};
        Queue<int[]> q = new Queue<int[]>();
        for(int i =0; i<m; i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j]=='1')
                {
                    count++;
                    q.Enqueue(new int[]{i,j});
                    grid[i][j] = '0';
                    while(q.Count!=0)
                    {
                        int[] curr = q.Dequeue();
                        foreach(int[] dir in dirs)
                        {
                            int r = curr[0]+ dir[0];
                            int c = curr[1]+ dir[1];
                            if(r<m && c<n && r>=0 && c>=0 && grid[r][c]=='1')
                            {
                                q.Enqueue(new int[]{r,c});
                                grid[r][c]='0';
                            }
                        }
                    }
                }
            }
        }
        return count;
    }
}

File: Sample.java
// Time Complexity : O(Length * product of all integers: K)
// Space Complexity :O(Length * product of all integers: K)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
//DFS
class Solution {
    int i =0;
    public String decodeString(String s) {
        int currNum = 0;
        StringBuilder currStr = new StringBuilder();
        while(i < s.length())
        {
            char c = s.charAt(i);
            i++;
            if(Character.isDigit(c))
            {
                currNum = currNum*10+c-'0';
            }
            else if(c=='[')
            {
                String baby = decodeString(s);
                for(int k = 0;k<currNum;k++)
                {
                    currStr.append(baby);
                }
                currNum=0;
            }
            else if(c==']')
            {
                return currStr.toString();
            }
            else
            {
                currStr.append(c);
            }
        }
        return currStr.toString();
    }
}

// Time Complexity : O(m*n)
// Space Complexity :O(m*n) for DFS (it was min(m,n) in case BFS
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : yes. syntax error plus logical error
//DFS
class Solution {
    int m;
    int n;
    int [][] dirs;
    public int numIslands(char[][] grid) {
        int count = 0;
        m = grid.length;
        n = grid[0].length;
        dirs = new int[][]{{-1,0},{0,1},{1,0},{0,-1}};
        for(int i=0;i<m;i++)
        {
            for(int j=0;j<n;j++)
            {
                if(grid[i][j]=='1')
                {
                    count++;
                    dfs(grid, i, j);
                }
            }
        }
        return count;
    }
    private void dfs(char[][] grid, int i,int j)
    {
        if(i<0||j<0||i>=m||j>=n||grid[i][j]!='1') return;
        grid[i][j] = '0';
        for(int[] dir:dirs)
        {
            int r = dir[0]+i;
            int c= dir[1]+j;
            dfs(grid, r, c);
        }
    }
}
